### ISSUE TITLE
spotifyd: crash on version 0.3.3

### ISSUE THREAD

**Issue by paralin** [Spotifyd/spotifyd#1033]
Using alsa backend on an arm64 machine (raspberry pi 4) with alsa. I've tested alsa and it works fine on its own. Spotifyd used to work here, too. But now it crashes when I select the speaker in the spotifyd app.

I found the error with --verbose:

```
kMessageTypePause "Pixel XL" 83311387d7a81a3ac100365756131ae41a4d6d66 1954805116 1640484825197 
layStatusPlay
PANIC: Shutting down spotifyd. Error message: called `Result::unwrap()` on an `Err` value: Send
ror(Pause)
drop Spirc[0]
```

I don't see anything else other than the panic unwrapping the Err value, presumably the actual error is that Err value but it's not being shown in the logs.

This is with latest HEAD - 603bcc4d802d2b8f45ac517c0513eb4096252d7b - and using alsa backend (no other backend is available and pulseaudio is not running) with the command line arguments:

```
  --no-daemon \
  --device-name "My Speaker" --bitrate 320 \
  -c /data/cache --device-type speaker \
  --username "myusername" --password "mypassword" \
  --verbose
```

I also tried with minimal arguments - --verbose only for example. Sometimes I can select the speaker successfully, but then it crashes when I try to play something.

Thanks for any help.

**Comment by paralin** [comment]
It doesn't always print the error - sometimes just this:

```
received packet from 192.168.1.14:5353
received packet from 192.168.1.14:5353 with no query
registering event source with poller: token=Token(33554437), interests=READABLE | WRITABLE
deregistering event source from poller
registering event source with poller: token=Token(50331653), interests=READABLE | WRITABLE
POST "/" {}
registering event source with poller: token=Token(6), interests=READABLE | WRITABLE
deregistering event source from poller
received packet from 192.168.1.13:5353
received question: IN _spotify-connect._tcp.local
found interface Interface { name: "wlan0", addr: V4(Ifv4Addr { ip: 192.168.1.14, netmask: 255.2
.255.0, broadcast: Some(192.168.1.255) }) }
found interface Interface { name: "docker_gwbridge", addr: V4(Ifv4Addr { ip: 172.18.0.1, netmas
 255.255.0.0, broadcast: Some(172.18.255.255) }) }
found interface Interface { name: "docker0", addr: V4(Ifv4Addr { ip: 172.17.0.1, netmask: 255.2
.0.0, broadcast: Some(172.17.255.255) }) }
found interface Interface { name: "tailscale0", addr: V4(Ifv4Addr { ip: 100.64.0.11, netmask: 2
.255.255.255, broadcast: None }) }
sending packet to 224.0.0.251:5353
received packet from 192.168.1.14:5353
received packet from 192.168.1.14:5353 with no query
Connecting to AP "ap.spotify.com:443"
registering event source with poller: token=Token(16777222), interests=READABLE | WRITABLE
flushing framed transport
writing; remaining=374
framed transport flushed
deregistering event source from poller
flushing framed transport
framed transport flushed
attempting to decode a frame
frame decoded from buffer
attempting to decode a frame
drop Spirc[0]
Shutting down player thread ...
drop PlayerInternal[0]
PlayerInternal thread finished.
drop Session[0]
drop MercuryManager
flushing framed transport
framed transport flushed
drop Dispatch
deregistering event source from poller
deregistering event source from poller
deregistering event source from poller
Shutting down discovery server
deregistering event source from poller
deregistering event source from poller

```

**Comment by eladyn** [comment]
Hi and thanks for the report!

I'm currently running `spotifyd` (v0.3.3) on my RPi 4 just fine (but with armhf), so there shouldn't be general problems with the current release. Could you share some more information about your situation?

- What OS are you using? (`lsb_release -a`)
- Did you build the binary yourself or did you use a prebuilt one from the releases?
- If the former, what features did you enable. If the latter, what binary did you try?

Also, if you get to reproduce the error, could you please paste the full log output as especially the first one does not really look complete? (I may be wrong on this one)

**Comment by paralin** [comment]
@eladyn using SkiffOS 2021.11.2. (Buildroot 2021.11). Built all binaries myself - spotifyd was built in an Alpine linux container (musl c which might be the reason) with "cargo build". 

I switched to using librespot and it's working fine. If this is not a reproducible crash, feel free to close. If I run into it again I'll post more info.

**Comment by eladyn** [comment]
Wow, that sounds like an interesting setup! However, I don't think that this causes the problem, especially since librespot is working fine for you, but it is probably the easiest to wait for your (or someone else's) complete logs before making any hypothesis.

*(Btw, I'm not a maintainer, just another user trying to help out, so I won't be able to close the issue :smile:)*

**Comment by agneevX** [comment]
>   `--device-name "My Speaker"`

Could it be because spaces aren't supported in device name?

**Comment by eladyn** [comment]
In theory, the example config states indeed that

> # The name that gets displayed under the connect tab on
> # official clients. Spaces are not allowed!

(as you said.)

However, the configuration / cli option parsing code provides [a fallback](https://github.com/Spotifyd/spotifyd/blob/603bcc4d802d2b8f45ac517c0513eb4096252d7b/src/config.rs#L627-L632) to this, so this shouldn't be the problem. Indeed, it just ignores my specified name, if I use "My Device" and instead it uses `Spotifyd@<hostname>`.

**Comment by agneevX** [comment]
On a side note, librespot supports spaces, so I'm confused as to why spotifyd doesn't.

**Comment by paralin** [comment]
@eladyn I'd love to add a apps/librespot and/or apps/spotifyd to Skiff, either through a Dockerfile + container image pull or actually adding the Rust packages to Buildroot directly

**Comment by eladyn** [comment]
Ok, so I now got a setup that _should_ be quite similar to yours to work (finally). I'm not sure wether I encountered any of the errors you did:

1. After some problems, I managed to get SkiffOS from the [latest](https://github.com/skiffos/SkiffOS/commit/7bd4d79a7e0d2da002249a5b74b45705789165ef) commit with `SKIFF_CONFIG=virt/qemu,core/alpine` to work. (I had to increase the size of the persistent storage to be able to compile `spotifyd` and I needed to add an audio device of course.)
3. In the Alpine container, I installed the necessary packages (`apk add rustup alsa-utils alsa-lib alsa-lib-dev openssl-dev build-base`).
4. Then I compiled `spotifyd` on [version v0.3.3](https://github.com/Spotifyd/spotifyd/releases/tag/v0.3.3) with the defaults (just `cargo build`). I had to add the following contents to cargo's `config.toml`:
   ```toml
   [build]
   rustflags = [ "-C", "target-feature=-crt-static" ]
   ```
   since I ran into [this](https://github.com/diwic/alsa-rs/issues/59) problem.

Afterwards, running the compiled `spotifyd` worked fine and produced correct audio output.

```console
~/spotifyd-0.3.3 # sudo target/debug/spotifyd --username "<username>" --password '<password>' --device-name "SkiffOS" --no-daemon
No config file specified. Running with default values
No proxy specified
Using software volume controller.
Connecting to AP "ap.spotify.com:443"
Authenticated as "<user>" !
Country: "DE"
Using Alsa sink with format: S16
Loading <...> with Spotify URI <spotify:track:...>
Vorbis error: VorbisError(Initial file headers are corrupt)
<...> (200000 ms) loaded
Loading <...> with Spotify URI <spotify:track:...>
<...> (220000 ms) loaded
```

Maybe this helps you, to get your problem solved? Otherwise, as I said, a complete verbose log from start to end would be good. :smiley:

**Comment by paralin** [comment]
@eladyn Wow, respect, you actually went and tested it 💯 

I think the problem probably was related to compiling on the pi4 64 bit. I'll go test it again this weekend and post here if it's still busted.

**Comment by Freyert** [comment]
@paralin, if you are planning to compile `spotifyd` it may be better to retrieve a stacktrace first. It will be a little more clunky to get your `spotifyd` client running, but will capture some very valuable data.

Here's what I want you to do. Instead of running your spotifyd as you normally do; run it like so:

    git clone git@github.com:Spotifyd/spotifyd.git; git checkout v0.3.3
    RUST_BACKTRACE=1 cargo run

Now when it panics we should get a backtrace and can identify exactly where the panic originates from.

http://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/second-edition/ch09-01-unrecoverable-errors-with-panic.html

**Comment by paralin** [comment]
@Freyert Will try, thanks. I haven't had access to the speaker in a while, and they're using librespot in the meantime. Will test when I can!

**Comment by Freyert** [comment]
Absolutely! Would be very interested to hear about your speaker set up. I don't know much about speakers, but I got a little excited about using a raspberry PI to drive some old, but nice speakers.

**Comment by paralin** [comment]
It's a Vizio sound bar with optical audio using a hifiberry digi pro on a
pi 4. Using SkiffOS and either librespot or spotifyd in a container

**Comment by 3579km** [comment]
This bug is actual on my device

Should we report on https://github.com/Spotifyd/spotifyd?

**Comment by eladyn** [comment]
I'm going to close this, since `v0.3.3` is a little outdated by now and there hasn't been anything new here. Feel free to reopen however!
